### PR TITLE
Adopt native complex dtype in griffnlim

### PR DESCRIPTION
**Pull request by mthrok**
Get rid of pseudo complex type in `F.griffinlim`.

Part of #1337 

`griffinlim` is tested
 - ✅ TorchScript on CPU/CUDA for `float32` and `float64`
 - ✅ `librosa` consistency on CPU
 - ✅ Batch consistency on CPU

**TODO**
 - [x] Autograd test (updated from #1421)
 - [x] Benchmark
 - [x] Remove the hardcoded `cfloat` in the intermediate variable.

## Benchmark

<details><summary>script</summary>

```bash
#!/usr/bin/env bash

OMP_NUM_THREADS=1 numactl --membind 0 --cpubind 0 python -m timeit -s """
import torch
import torchaudio;

batch_size = 32
num_channels = 1
num_freq = 201
num_frames = 400

torch.manual_seed(0);
spec = torch.randn(batch_size, num_channels, num_freq, num_frames, dtype=torch.float32, device='cpu');
griffinlim = torchaudio.transforms.GriffinLim()
""" """
griffinlim(spec)
"""
```

</details>

| device |           master (8d2eeb1)           |                This PR                |
|:------:|:------------------------------------:|:-------------------------------------:|
|    CPU | 1 loop, best of 5: 6.48 sec per loop | 1 loop, best of 5: 3.57 sec per loop  |
|    GPU | 1 loop, best of 5: 128 msec per loop | 1 loop, best of 5: 93.4 msec per loop |

### CODE REVIEW

File: torchaudio/functional/functional.py
Comment (mthrok) [inline torchaudio/functional/functional.py:183]: @anjali411 Do you know an elegant and TorchScript-compatible way to initialize Tensors with value `1+0i`?

Comment (anjali411) [reply torchaudio/functional/functional.py:183]: After #52881 is merged, `torch.full(batch, freq, frames, 1j, dtype=torch.cfloat, device=specgram.device)` should work

Comment (mthrok) [reply torchaudio/functional/functional.py]: it worked locally. pushing the change.

File: torchaudio/functional/functional.py
Comment (mthrok) [inline torchaudio/functional/functional.py:180]: TODO: check to see if `cfloat` is preferred.

Comment (anjali411) [reply torchaudio/functional/functional.py:180]: yeah or even take a step back and consider why we were creating float tensors by default and if it makes sense to  add a dtype `kwarg` that defaults to float

**Review by anjali411 — COMMENTED** [review]
needs JIT and autograd tests to verify if everything works as expected

File: torchaudio/functional/functional.py
Comment (mthrok) [reply torchaudio/functional/functional.py:173]: @anjali411

I looked into this and the followings are my finding;

1. The use of `cfloat` here is fine when the input Tensor is `float32` and `float64`.
This is because the dtype of `angles` are updated to the correct one in the next `for` loop. Specifically, `specgram * angles` there produces the same dtype as `specgram` so after the first iteration, all the Tensor has the matching dtype as the input.
2. The above suggests that the `griffinlim` will not work for `float16` input type. But I did test and `pow` is not implemented for `float16`. (see the detail bellow) so I do not think it's and issue.

With this finding, I suggest to proceed with the current approach (using `cfloat` for intermediate variables) since it is working for `float32` and `float64`. Also note that `float16` is not in general tested in `torchaudio`.

---

The following is the result of running `griffinlim` for the different dtype on before/after this PR.

### Script
<Details><summary>script</summary>

```python
import torch
import torchaudio.transforms as T


def test(waveform, dtype):
    n_fft = 1024
    win_length = None
    hop_length = 512

    spectrogram = T.Spectrogram(
        n_fft=n_fft,
        win_length=win_length,
        hop_length=hop_length,
    )
    griffin_lim = T.GriffinLim(
        n_fft=n_fft,
        win_length=win_length,
        hop_length=hop_length,
    )

    spec = spectrogram(waveform).to(dtype)
    print('*******')
    print('input:', spec.dtype)
    recon = griffin_lim(spec)
    print('output:', recon.dtype)
    assert recon.dtype == spec.dtype
    return recon


torch.random.manual_seed(0)
waveform = torch.randn(2, 1024)


test(waveform, torch.float64)
test(waveform, torch.float32)
test(waveform, torch.float16)
```

<Details>

### Result
#### Before applying this PR (commit 0433b7aa)

```
*******
input: torch.float64
/home/moto/conda/envs/PY3.8-cuda101/lib/python3.8/site-packages/torch/functional.py:655: UserWarning: istft will require a complex-valued input tensor in a future PyTorch release. Matching the output from stft with return_complex=True.  (Triggered internally at  /opt/conda/conda-bld/pytorch_1617001512472/work/aten/src/ATen/native/SpectralOps.cpp:807.)
  return _VF.istft(input, n_fft, hop_length, win_length, window, center,  # type: ignore
output: torch.float64
*******
input: torch.float32
output: torch.float32
*******
input: torch.float16
Traceback (most recent call last):
  File "foo.py", line 36, in <module>
    test(waveform, torch.float16)
  File "foo.py", line 24, in test
    recon = griffin_lim(spec)
  File "/home/moto/conda/envs/PY3.8-cuda101/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1015, in _call_impl
    return forward_call(*input, **kwargs)
  File "/scratch/moto/torchaudio/torchaudio/transforms.py", line 189, in forward
    return F.griffinlim(specgram, self.window, self.n_fft, self.hop_length, self.win_length, self.power,
  File "/scratch/moto/torchaudio/torchaudio/functional/functional.py", line 168, in griffinlim
    specgram = specgram.pow(1 / power)
RuntimeError: "pow" not implemented for 'Half'
```

#### After applying this PR

```
*******
input: torch.float64
output: torch.float64
*******
input: torch.float32
output: torch.float32
*******
input: torch.float16
Traceback (most recent call last):
  File "foo.py", line 36, in <module>
    test(waveform, torch.float16)
  File "foo.py", line 24, in test
    recon = griffin_lim(spec)
  File "/home/moto/conda/envs/PY3.8-cuda101/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1015, in _call_impl
    return forward_call(*input, **kwargs)
  File "/scratch/moto/torchaudio/torchaudio/transforms.py", line 189, in forward
    return F.griffinlim(specgram, self.window, self.n_fft, self.hop_length, self.win_length, self.power,
  File "/scratch/moto/torchaudio/torchaudio/functional/functional.py", line 168, in griffinlim
    specgram = specgram.pow(1 / power)
RuntimeError: "pow" not implemented for 'Half'
```

Comment (anjali411) [reply torchaudio/functional/functional.py:186]: This PR https://github.com/pytorch/pytorch/pull/50999 recently added `pow` for `torch.float16`

Comment (anjali411) [reply torchaudio/functional/functional.py:186]: > The use of cfloat here is fine when the input Tensor is float32 and float64.
This is because the dtype of angles are updated to the correct one in the next for loop. Specifically, specgram * angles there produces the same dtype as specgram so after the first iteration, all the Tensor has the matching dtype as the input.

So in that case, I think it makes sense to instead create angles of the same dtype as `specgram` since multiplication in the for loop (`specgram` * `angles`) [promotes all input tensors to the same dtype](https://github.com/pytorch/pytorch/blob/a102adb55e00ea3bc567f6a700d19895f99e148a/aten/src/ATen/TensorIterator.h#L525-L531), so I think it makes sense to create `angles` with dtype `specgram.dtype`.

**Comment by anjali411** [comment]
@mthrok this PR needs performance benchmarking!

File: torchaudio/functional/functional.py
Comment (mthrok) [reply torchaudio/functional/functional.py:186]: oh this could be the reason why autograd test in #1421 is failing.

Comment (mthrok) [reply torchaudio/functional/functional.py]: > This PR [pytorch/pytorch#50999](https://github.com/pytorch/pytorch/pull/50999) recently added `pow` for `torch.float16`

Looks like it's still landing? https://github.com/pytorch/pytorch/pull/55280



> So in that case, I think it makes sense to instead create angles of the same dtype as `specgram` since multiplication in the for loop (`specgram` * `angles`) [promotes all input tensors to the same dtype](https://github.com/pytorch/pytorch/blob/a102adb55e00ea3bc567f6a700d19895f99e148a/aten/src/ATen/TensorIterator.h#L525-L531), so I think it makes sense to create `angles` with dtype `specgram.dtype`.

Added the necessary change.

**Review by anjali411 — APPROVED** [review]
LGTM! Thanks Moto. let's merge this after #1421 is merged

File: torchaudio/functional/functional.py
Comment (anjali411) [inline torchaudio/functional/functional.py:195]: wait the angles here should be a floating point tensor, no?

Comment (mthrok) [reply torchaudio/functional/functional.py:195]: No, `angles` is complex value. `Griffin-Lim` algorithm iteratively optimizes the phase (or the direction in complex plain) of each element of the given spectrogram so that at the end `istft` give the original waveform.

Comment (mthrok) [reply torchaudio/functional/functional.py:195]: *The original implementation was constructing the complex value Tensor called `angles` from real valued tensor called `angle` and `magnitude`.

File: torchaudio/functional/functional.py
Comment (imaginary-person) [reply torchaudio/functional/functional.py]: Sorry, #50999 would hopefully reland soon via #55280, and it'd enable `pow` for `float16` on CPU.

`pow` for `float16` is already supported on CUDA.

Comment (mthrok) [reply torchaudio/functional/functional.py]: @imaginary-person Thanks for the info.

**Comment by mthrok** [comment]
@anjali411 

The autograd tests are failing for the case `rand_init=True`. However, it does not fail on my local env with PyTorch nightly from March 29th.

```
[conda] pytorch                   1.9.0.dev20210329 py3.8_cuda10.1_cudnn7.6.3_0    pytorch-nightly
```

Failing test log: https://app.circleci.com/pipelines/github/pytorch/audio/5665/workflows/746a53f5-6f18-4f2e-b717-869bf342eaf4/jobs/194528

I am wondering if this is a regression. What do you think? The following is the corresponding code;

https://github.com/mthrok/audio/blob/17f61e73f8f4b0f314bb7356aa3a2222c4038347/torchaudio/functional/functional.py#L193-L201

**Comment by yoyololicon** [comment]
> @anjali411
> 
> The autograd tests are failing for the case `rand_init=True`.

@mthrok 

Hmm, I think it's an expected behavior, cuz numerical method will (I'm not sure though) forward the function multiple times, and with `rand_init=True`, each forward pass will start with different initial phases, result in different output.

For functions like it which have random behavior internally, maybe a `.backward()` at the end of test is enough, to make sure that gradients can be propogated without problems.

**Comment by mthrok** [comment]
> > @anjali411
> > The autograd tests are failing for the case `rand_init=True`.
> 
> @mthrok
> 
> Hmm, I think it's an expected behavior, cuz numerical method will (I'm not sure though) forward the function multiple times, and with `rand_init=True`, each forward pass will start with different initial phases, result in different output.
> 
> For functions like it which have random behavior internally, maybe a `.backward()` at the end of test is enough, to make sure that gradients can be propogated without problems.

Hi @yoyololicon 

Thanks for the insight. You are right. I added a wrapper that sets the random seed and the test passes now.
(I do not know why I thought it was working on my local env...)

File: test/torchaudio_unittest/transforms/autograd_test_impl.py
Comment (anjali411) [inline test/torchaudio_unittest/transforms/autograd_test_impl.py:24]:
```suggestion
        torch.random.manual_seed(self.seed)
```

**Comment by mthrok** [comment]
@anjali411 This PR is ready for the final review.

**Pull request merged**
